### ISSUE TITLE
`sfdx force:source:push` pushes all files if source dir is symlink (ubuntu 22.04)  

### ISSUE THREAD

**Issue by vytautasgermanavicius** [forcedotcom/cli#1851]
### Summary

`sfdx force:source:push` pushes all files if source dir is symlink .

OS: ubuntu 22.04

### Steps To Reproduce:

1. Make `force-app` to be a relative symlink to an another dir 
1. `sfdx force:source:push` 
1. Modify 1 file
1. sfdx force:source:push`

### Expected result

First push pushes 2542 files.
Second push pushes 1 file.

### Actual result

First `push` pushes 2542 files.
Second `push` pushes 2542 files.

### System Information

OS: ubuntu 22.04
bash

$ sfdx -v
sfdx-cli/7.181.1 linux-x64 node-v18.12.1

**Comment by armaid** [comment]
I also have the same problem.
OS: Windows 10
SFDX version: sfdx-cli/7.178.0 win32-x64 node-v17.8.0

**Comment by mshanemc** [comment]
> tldr; don't use symlinks

I've been tinkering with this.  Here's what I've found:

`git` treats a symlink like a file.  For example, `git status` won't show all the files in the symlinked directory.
the CLI uses git to manage tracking of local changes
there's a bug in the git implementation we're using that makes it follow symlinks when it shouldn't...for example, when running, `git status` to ask what to push

So the net result is that that `push` should have never deployed anything at all.  `force:source:status` shouldn't show files inside a symlinked directory.  When the bug is fixed, it won't.

I'd encourage you not to rely on symlink working in sfdx projects if you're using source tracking.

**Comment by vytautasgermanavicius** [comment]
@mshanemc Symlinks are essential part on complicated projects, which uses multiple git repositories to construct a system. 

You assume, that 1 org  equals 1 git repository, but this valid only for small projects. 

Sysmlinks are way to include  3rd party code into the project without making mess in git.

**Comment by vytautasgermanavicius** [comment]
@mshanemc Should I register feature request to be able to use symlinks with push/pull commands?

Use cases:   

* Company have project, which uses 42 external git repositories, like  nebula , amoss, fflib, Sobjectfabricator and etc. Some of external libraries contains  functiopnality or files we do not want to deploy to prod org, like examples of usage, not deployable files, bash scripts, documentation end etc.  So symlinks are used to refer directories Company want to have in prod org. 

* Company have very complicated system - 1500 classes. I want to test part of functionality in scratch org without making clone of the prod org. I have scratch org repository, to construcct scratch org, which contains scratch org specific data, scripts, test data, which is not part of prod org, and symlinks to folders with classes other metadata i want to use in my scratch org. 

* Company outsource part of the system. Company needs to make sure, that nobody can get complete system before it is built in preprod , or mater scratch. Eanch part is separate repository, which contains metadata whish are shared with master syste, and part which is not shared. changes can go both ways: from master to parts, and from parts to master.  This can be easily managed with symlinks, but makes synchronization hell with copying files. 

These are work patterns which can be easily in any other programming language , virtualization solution, and ERP system.

**Comment by mshanemc** [comment]
> You assume, that 1 org equals 1 git repository, but this valid only for small projects.

There's lots of people using multiple git repositories (and multiple sfdx-projects) to manage different "areas" of an org.  With symlinks, you're putting them all into 1 sfdx-project and 1 repository.

There's not an easy way to make this work the way you want it to because of how `git` works.  

Can you elaborate on why symlinks are used instead of git submodules?

**Comment by vytautasgermanavicius** [comment]
> Can you elaborate on why symlinks are used instead of git submodules?

Purposes of submodules and symlinks are different.

Example of folder structure:

Libraries
 \-–> Git_repository1 
    \---> scratch specific metadata 1 
    \--> Prod specific metadata 1
  \---> Git repository 2
    \---> Feature 21
    \---> Feature 22
SandboxOrg1
  \---> symlink to  Prod specific metadata 1
  \----> symlink to  Feature 22 
  \----> org specific metadata for Feature 22
  
Scratch org1   
    \--->  symlink to scratch specific metadata 1 
    \-->  symlink to Prod specific metadata 1
    \--->  symlink to Feature 21
    \--->  symlink to Feature 22
    \----> scratchorg specific metadata for Feature 22

So, when i  do  sfdx push in SandboxOrg1 dir, i do not want to push Feature 21 , when i do sfdx pull i expect thet ofg specific metadata will remain in SandboxOrg1 dir, but changes in Feature 22 will go to lib an will be shared with Scratch org1.
